File: c++/60_Permutation_Sequence.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    string getPermutation(int n, int k) {
    	int factorial = 1;
    	for (int i = 1; i < n; i++)
    		factorial *= i;
    	k--;
    	int sets[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    	vector<int> setNum(sets, sets + 9);
    	string res;
    	int tag = 0;
    	while (k >= 0 && tag < n) {
    		tag++;
    		int quotient = k / factorial;
    		k -= quotient * factorial;
    		res.push_back('0' + setNum[quotient]);
    		setNum.erase(setNum.begin() + quotient);
    		if (n == tag)
    			factorial = 1;
    		else
    			factorial /= (n - tag);
    	}
    	return res;
    }
};

int main() {
	Solution s;
	cout << s.getPermutation(2, 1) << endl;

}

File: python/60_Permutation_Sequence.py
#!/usr/bin/python


class Solution(object):
    def getPermutation(self, n, k):
    	factorial = 1
    	for i in range(1, n):
    		factorial *= i
    	if k > factorial * n:
    		return -1
    	k -= 1
    	sets = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    	tag, res = 0, ""
    	
        while k >= 0 and tag < n:
        	tag += 1
        	quotient = k / factorial
        	k -= quotient * factorial
        	res += str(sets[quotient])
        	sets.remove(sets[quotient])
        	if n == tag:
        		factorial = 1
        	else:
        		factorial /= (n - tag)
        return res
        

def main():
    s = Solution()
    print s.getPermutation(4, 9)

if __name__ == "__main__":
    main()